compiler/crates/relay-typegen/tests/generate_flow/fixtures/typename-on-union.expected

```diff
@@ -80,22 +80,346 @@ fragment TypenameAliases on Actor {
+import type { FragmentReference } from "relay-runtime";
+declare export opaque type TypenameWithCommonSelections$ref: FragmentReference;
+declare export opaque type TypenameWithCommonSelections$fragmentType: TypenameWithCommonSelections$ref;
+export type TypenameWithCommonSelections = {|
+  +__typename: "User",
+  +firstName: ?string,
+|} | {|
+  +__typename: "Page",
+  +username: ?string,
+|} & {|
+  +name: ?string,
+  +$refType: TypenameWithCommonSelections$ref,
+|};
+export type TypenameWithCommonSelections$data = TypenameWithCommonSelections;
+export type TypenameWithCommonSelections$key = {
+  +$data?: TypenameWithCommonSelections$data,
+  +$fragmentRefs: TypenameWithCommonSelections$ref,
+  ...
+};
+-------------------------------------------------------------------------------
+import type { FragmentReference } from "relay-runtime";
+declare export opaque type TypenameWithoutSpreads$ref: FragmentReference;
+declare export opaque type TypenameWithoutSpreads$fragmentType: TypenameWithoutSpreads$ref;
+export type TypenameWithoutSpreads = {|
+  +__typename: "User",
+  +firstName: ?string,
+  +$refType: TypenameWithoutSpreads$ref,
+|};
+export type TypenameWithoutSpreads$data = TypenameWithoutSpreads;
+export type TypenameWithoutSpreads$key = {
+  +$data?: TypenameWithoutSpreads$data,
+  +$fragmentRefs: TypenameWithoutSpreads$ref,
+  ...
+};
+-------------------------------------------------------------------------------
+import type { FragmentReference } from "relay-runtime";
+declare export opaque type TypenameWithoutSpreadsAbstractType$ref: FragmentReference;
+declare export opaque type TypenameWithoutSpreadsAbstractType$fragmentType: TypenameWithoutSpreadsAbstractType$ref;
+export type TypenameWithoutSpreadsAbstractType = {|
+  +__typename: string,
+  +id: string,
+  +$refType: TypenameWithoutSpreadsAbstractType$ref,
+|};
+export type TypenameWithoutSpreadsAbstractType$data = TypenameWithoutSpreadsAbstractType;
+export type TypenameWithoutSpreadsAbstractType$key = {
+  +$data?: TypenameWithoutSpreadsAbstractType$data,
+  +$fragmentRefs: TypenameWithoutSpreadsAbstractType$ref,
+  ...
+};
+-------------------------------------------------------------------------------
+/*
+ * future proof enums: false
+ * future proof abstract types: true
+ */
+
 import type { FragmentReference } from "relay-runtime";
 declare export opaque type TypenameAlias$ref: FragmentReference;
 declare export opaque type TypenameAlias$fragmentType: TypenameAlias$ref;
 export type TypenameAlias = {|
   +_typeAlias: "User",
   +firstName: ?string,
-  +$refType: TypenameAlias$ref,
 |} | {|
   +_typeAlias: "Page",
   +username: ?string,
-  +$refType: TypenameAlias$ref,
 |} | {|
   // This will never be '%other', but we need some
   // value in case none of the concrete values match.
-  +_typeAlias: "%other",
-  +$refType: TypenameAlias$ref,
+  +_typeAlias: "%other"
+|} & {|
+  +$refType: TypenameAlias$ref
 |};
```

Comment:
Unfortunately, this doesn't work with flow, while it appears to work with Typescript.

What `{| a: number |} & {| b: number |}` expresses in Flow is: an object with exactly key `a`, nothing more and at the same time and object with exactly the key b and nothing more. That object can obviously not exist, so access to properties fails. I'm not sure what TS makes of this, not a type system expert.

- [flow example](https://flow.org/try/#0C4TwDgpgBAKuEDsCGBbCB1AlsAFgYQHsUUCEBlCAGwgGNhNSBnKAXigG8AfAKCigGoA+oNCRkaAFxQARAFVGEAE7SANLwEAzTIsbAAcqghSA-LsWYEAczWcAvlE4cefISPjijMgApJLEVer8AK4Kih4mZhbW3HZQAGROgeFQpsDmVja2ANzc3BpBCHQMCFAaABQAHlJwYoZYuITEpBTURUwAlBzqmBpQlQB0wqKIhqxsAOTySuOd7Op8Ff1aOvqG6rbctkA)
- [typescript example](https://www.typescriptlang.org/play?#code/C4TwDgpgBAKuEDsCGBbCB1AlsAFgYQHsUUCEBlCAGwgGNhNSBnKAXigAoBvAKCigCcISACalKIKAH1JoSMjQAuKACIAqowj9lAGl4ChohOKgAzTP0bAAcqghLL-TAgDmUAD5QEAV0qVdAX3coHj5BETEJaVlEWyVlAAUkZwgdPTDDYy8Nfnk7KAcnVw9vXwCASigAMmC0gwjPWPzgRxcgkr9ufwBubm4TLwQ6BgRTdgAPJTg5WyxcQmJSCmohpgqQqEwTDjGAOij4XNY2AHJ1TWO1vT5dswtrWz1-TqA)

Reply:
TypeScript doesn't have this concept of "exact" objects, so it's just a regular intersection. I made some changes so typegen writers that support exact objects get fed a regular union of exact objects, with the shared properties being duplicated in each union member. For TypeScript (which does not support the exact objects) or any other future languages that do not support exact objects, it still issues an intersection AST.